File: go/isPalindrome.go
// https://leetcode.com/problems/valid-palindrome

package leetcode

import (
	"unicode"
)

func isAlNum(r rune) bool {
	return unicode.IsLetter(r) || unicode.IsNumber(r)
}

func isPalindrome(s string) bool {
	left := 0
	right := len(s) - 1
	runes := []rune(s)

	for left < right {
		if !isAlNum(runes[left]) {
			left += 1
			continue
		}

		if !isAlNum(runes[right]) {
			right -= 1
			continue
		}

		if unicode.ToLower(runes[left]) != unicode.ToLower(runes[right]) {
			return false
		}

		left += 1
		right -= 1
	}

	return true
}

File: rust/src/is_palindrome.rs
// https://leetcode.com/problems/valid-palindrome
use crate::leetcode::Solution;

impl Solution {
    pub fn is_palindrome(s: String) -> bool {
        let chars: Vec<char> = s
            .chars()
            .filter(|c| c.is_alphanumeric())
            .map(|c| c.to_lowercase().next().unwrap())
            .collect();

        let mut left = 0;
        let mut right = if chars.is_empty() { 0 } else { chars.len() - 1 };
        while left < right {
            if chars[left] != chars[right] {
                return false;
            }
            left += 1;
            right -= 1;
        }

        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example_1() {
        let got = Solution::is_palindrome("A man, a plan, a canal: Panama".to_string());
        assert_eq!(got, true);
    }

    #[test]
    fn example_2() {
        let got = Solution::is_palindrome("race a car".to_string());
        assert_eq!(got, false);
    }

    #[test]
    fn example_3() {
        let got = Solution::is_palindrome(" ".to_string());
        assert_eq!(got, true);
    }
}
